TLSv1.3: Add HelloRetryRequest support

mattcaswell opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] tests are added or updated

##### Description of change
<!-- Provide a description of the changes.

If it fixes a github issue, add Fixes #XXXX.
-->
This PR adds support for HelloRetryRequest messages.

It is marked as WIP because I still need to update the tests.

mattcaswell commented:
@davidben - while testing this I noticed something unexpected with the sandbox site. At one point during dev I was sending an invalid binder. However sandbox *successfully* created a connection (non resumed). I would have expected it to abort the handshake when it hit the invalid binder:

"Prior to accepting PSK key establishment, the server MUST validate the corresponding binder value (see Section 4.2.6.1 below).  If this value is not present or does not validate, the server MUST abort the handshake."

Is this a bug? Intentional? Or a misunderstanding on my part?

kaduk commented:
It might be a misunderstanding if you were also sending something else that would cause the server to not want to resume, since the server is only supposed to check the one binder it might want to use; it's not supposed to check all of them.  If it doesn't want to try to use any of them, then it doesn't need to check any of them.

mattcaswell commented:
Hmmm - maybe - although I subsequently tried it again with a deliberately corrupt binder, followed by the exact same test with a correct binder. With the correct binder it resumed with the corrupt binder it did not resume but did connect.

kaduk commented:
That sounds more like a bug with the sandbox site, yeah.  But we should let David chime in...

FdaSilvaYY commented on ssl/statem/extensions_clnt.c:
```diff
@@ -132,7 +132,8 @@ int tls_construct_ctos_ec_pt_formats(SSL *s, WPACKET *pkt, X509 *x,
 
-int tls_construct_ctos_supported_groups(SSL *s, WPACKET *pkt, X509 *x,
+int tls_construct_ctos_supported_groups(SSL *s, WPACKET *pkt,
+                                        unsigned int context,X509 *x,
```
Style: unsigned int context, X509 *x,

mattcaswell replied:
Fixed

FdaSilvaYY commented on ssl/statem/extensions_clnt.c:
```diff
@@ -524,12 +528,55 @@ int tls_construct_ctos_psk_kex_modes(SSL *s, WPACKET *pkt, X509 *x,
+static int add_key_share(SSL *s, WPACKET *pkt, unsigned int curve_id)
+{
+    unsigned char *encodedPoint = NULL;
+    EVP_PKEY *key_share_key = NULL;
```
useless init

mattcaswell replied:
Yeah, I know. I've been bitten enough times by travis complaining about an uninit variable (whether justly or not) that I tend to init anyway most often.

FdaSilvaYY commented on ssl/statem/extensions_clnt.c:
```diff
@@ -1179,6 +1206,49 @@ int tls_parse_stoc_key_share(SSL *s, PACKET *pkt, X509 *x, size_t chainidx,
     }
 
+    if ((context & EXT_TLS1_3_HELLO_RETRY_REQUEST) != 0) {
+        unsigned const char *pcurves = NULL;
```
Useless init.

mattcaswell replied:
As above.

mattcaswell commented:
Various test cases added. Also added `SSL_trace()` support and rebased to pick up the latest changes in master.

Taking this out of WIP. This is ready for review.

richsalz commented on ssl/statem/extensions.c:
```diff
@@ -965,22 +968,102 @@ static int final_key_share(SSL *s, unsigned int context, int sent, int *al)
+            size_t num_curves, clnt_num_curves, i;
+            unsigned int group_id;
+
+            /* Check a shared group exists */
```
check *if* ?

mattcaswell replied:
Fixed

richsalz commented on ssl/statem/extensions.c:
```diff
@@ -965,22 +968,102 @@ static int final_key_share(SSL *s, unsigned int context, int sent, int *al)
+            /* Find the first group we allow that is also in client's list */
+            for (i = 0, pcurvestmp = pcurves; i < num_curves;
+                 i++, pcurvestmp += 2) {
+                group_id = pcurvestmp[0] << 8 | pcurvestmp[1];
```
i will say yet again how much i hate this construct.

richsalz commented on ssl/statem/extensions_clnt.c:
```diff
@@ -108,8 +108,8 @@ static int use_ecc(SSL *s)
 
-int tls_construct_ctos_ec_pt_formats(SSL *s, WPACKET *pkt, X509 *x,
-                                     size_t chainidx, int *al)
+int tls_construct_ctos_ec_pt_formats(SSL *s, WPACKET *pkt, unsigned int context,
```
we now have SIX parameters on every parse function.  Time to make a custom struct that gets passed in I think.

mattcaswell replied:
Hmmm, I'm unconvinced. That would just introduce unnecessary complexity IMO, i.e. code to pack the struct prior to the call. It has no purpose other than to pass parameters to a function...which is what function parameters are for! I don't really see 6 parameters as a problem that needs solving.

richsalz replied:
Six lines of assignment in the common calling function vs pushing six values onto the stack every single time.  Not that this matters, it's such a micro-efficiency issue.

This is about code readability and maintainability.  Most functions don't need all parameters, and the #2270 (unused-arg) branch has many many ```(void)``` casts that wouldn't be needed.  As for maintainability, you've already added three parameters, and it's not clear you're done yet :)  And also for follow-on work, it's much less invasive to change one struct then to add a param everywhere.

mattcaswell replied:
Almost all functions will need at least 3 of those params. Hiding stuff like the SSL object and the WPACKET just doesn't seem helpful to me. I really do not see this as a readability gain - the opposite in fact...we will litter the code with lots of mystruct->pkt expressions or similar (or if not we have to have code to unpack the struct as well as pack it)

richsalz replied:
It's not hiding, no more than SSL hides things or WPACKET does.  I think it's better and more maintainable, and gave the best reasoning that I could.  Oh well.

levitte replied:
This dispute has helped me make my mind up regarding the usability of #2270.  Thank you.

richsalz commented on ssl/statem/extensions_clnt.c:
```diff
@@ -524,12 +528,57 @@ int tls_construct_ctos_psk_kex_modes(SSL *s, WPACKET *pkt, X509 *x,
+static int add_key_share(SSL *s, WPACKET *pkt, unsigned int curve_id)
+{
+    unsigned char *encodedPoint = NULL;
+    EVP_PKEY *key_share_key = NULL;
```
We should figure out style on initializing.  But since ```key_share_key``` is set in the very first statement of the function, and then encodedPoint, I would remove these.  Not going to argue about it strongly, however.  But, I do think encodedPoint is better written as ```encoded_point``` since we don't do camelCase.

mattcaswell replied:
Ok. Done.

richsalz commented on ssl/statem/extensions_clnt.c:
```diff
@@ -524,12 +528,57 @@ int tls_construct_ctos_psk_kex_modes(SSL *s, WPACKET *pkt, X509 *x,
+     * going to need to be able to save more than one EVP_PKEY. For now
+     * we reuse the existing tmp.pkey
+     */
+    s->s3->tmp.pkey = key_share_key;
```
tmp.pkey is always null?  no leak here?

mattcaswell replied:
There's an explicit test for it in `tls_construct_ctos_key_share()`, so there is no way it can be anything other than NULL here.

richsalz commented on ssl/statem/extensions_clnt.c:
```diff
@@ -1179,6 +1208,49 @@ int tls_parse_stoc_key_share(SSL *s, PACKET *pkt, X509 *x, size_t chainidx,
+            return 0;
+        }
+        for (i = 0; i < num_curves; i++, pcurves += 2) {
+            if (group_id == (unsigned int)((pcurves[0] << 8) | pcurves[1]))
```
maybe it's finally time for at least a macro b2c for bytestocurve?
```
#define b2c(p)          (unsigned int)((p[0]<<8)|p[1])
```

mattcaswell replied:
Done. Although I called it bytestogroup(). `b2c` to me just seems too brief and doesn't give you any indication what it does.

richsalz commented on ssl/statem/statem_clnt.c:
```diff
@@ -1432,6 +1463,52 @@ MSG_PROCESS_RETURN tls_process_server_hello(SSL *s, PACKET *pkt)
+static MSG_PROCESS_RETURN tls_process_hello_retry_request(SSL *s, PACKET *pkt)
+{
+    unsigned int sversion;
+    int protverr;
```
just call this errorcode?

mattcaswell replied:
Done

mattcaswell commented:
Updated with new commits to address the review feedback.

mattcaswell commented:
Added some commits to attempt to fix some Travis failures. Also rebased on top of latest master.

mattcaswell commented:
Travis failures appear to be unrelated to this PR and more to do with #2579.

Ping?

richsalz commented:
I just want to bring it up front so that it doesn't get lost in all the commits.  I think a custom struct is better than 3, no 6, no ...??? parameters to every function.

mattcaswell commented:
> I just want to bring it up front so that it doesn't get lost in all the commits. I think a custom struct is better than 3, no 6, no ...??? parameters to every function.

I'm assuming your intention was to draw @levitte's attention to this?

richsalz commented:
Not just him; I'd really like to see what others think -- maybe I'm out in the weeds on this.

levitte commented:
> I just want to bring it up front so that it doesn't get lost in all the commits. I think a custom struct is better than 3, no 6, no ...??? parameters to every function.

I both agree and disagree with you.  I agree *if* it's a tool to secure a stable (and a priori public) API while keeping the possible input to affected function flexible, that avoids having a gazillion different functions, one for each possible case.  For internal functions, I don't see the point unless there's a *huge* (6 isn't huge) number of parameters that are scarcelly used.

richsalz commented:
Most extension parsing doesn't need the cert/chain stuff.  Most SSL functions take an SSL object and one or two other parameters.  Put the parsing sub-struct inside the SSL object perhaps.

mattcaswell commented:
@richsalz how do we take this forward? I am keen to get this merged and have @levitte's approval. Do you object to the merge?

richsalz commented:
As I said off-line, I'm not going to minus-one this but I am unhappy.

mattcaswell commented:
Pushed. Thanks for the review everyone.
